src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,186 @@
+import type { Patch } from 'immer';
+
+import { BaseController } from '../BaseControllerV2';
+
+import type { RestrictedControllerMessenger } from '../ControllerMessenger';
+
+/**
+ * @type RateLimitState
+ * @property requests - Object containing number of requests in a given interval for each origin and api type combination
+ */
+export type RateLimitState<T extends string> = {
+  requests: Record<T, Record<string, number>>;
+};
+
+export interface CallArgs<T> {
+  /**
+   * Enum type to determine API type.
+   */
+  type: T;
+
+  /**
+   * Args passed to the API call
+   */
+  // @todo Type this?
+  args: any[];
```

Comment:
I think we just want to constrain things to an array here, so `unknown` should be fine. Perhaps args are optional, as well?:

```suggestion
  args?: unknown[];
```

Reply:
Done!